Collect number of chimeric reads

eitanbanks opened the pull request.

tfenne commented on src/java/picard/analysis/CollectWgsMetricsFromQuerySorted.java:
```diff
@@ -101,6 +104,12 @@
+         * The number of PF reads that map outside of a maximum insert size (100kb) or that have
+         * the two ends mapping to different chromosomes.
+         */
+        public double PF_CHIMERA_PAIRS;
```
`PF_CHIMERIC_PAIRS`?

tfenne commented on src/java/picard/analysis/CollectWgsMetricsFromQuerySorted.java:
```diff
@@ -270,6 +280,21 @@ private boolean isNoiseRead(final SAMRecord record) {
+     * @param minimumMappingQuality the minimum mapping quality
+     * @return true if the read pair is considered chimeric, false otherwise
+     */
+    private boolean isChimericReadPair(final QuerySortedReadPairIteratorUtil.ReadPair readPair, final int minimumMappingQuality) {
```
Since @nh13 just did something very similar in CollectAlignmentSummaryMetrics it would probably be good to extract this into a re-usable spot so that there's a reasonable fixed definition of Chimeras somewhere.  I could imagine:
```java
public class ChimeraUtil {
    public static int DEFAULT_INSERT_SIZE_LIMIT = 100000;
    public static int DEFAULT_MAPQ_LIMIT = 20;
    public Set<PairOrientation> DEFAULT_EXPECTED_ORIENTATIONS = EnumSet.of(PairOrientation.FR);

    public void isChimeric(final SAMRecord rec);
    public void isChimeric(final SAMRecord r1, final SAMRecord r2);
```
}

eitanbanks commented:
Good point.  Changed it.

ktibbett commented:
Can you please change all instances of "adaptor" to "adapter" to match the spelling we have used for this elsewhere in the code base?

eitanbanks commented:
@tfenne @ktibbett I've refactored things a bit and this PR is much larger than originally planned, so I squashed the commits and am requesting a new review from scratch if that's okay.  Does either of you want to do that?

tfenne commented:
I’m happy to take a pass through it, but probably won’t have time until tomorrow.

tfenne commented on src/java/picard/analysis/AdapterUtil.java:
```diff
@@ -0,0 +1,105 @@
+            IlluminaUtil.IlluminaAdapterPair.PAIRED_END.get3PrimeAdapter(),
+            IlluminaUtil.IlluminaAdapterPair.INDEXED.get5PrimeAdapter(),
+            IlluminaUtil.IlluminaAdapterPair.INDEXED.get3PrimeAdapter()
+    );
```
Perhaps it's time we added `DUAL_INDEXED` to this list?

eitanbanks replied:
Just spoke to @ktibbett.  I'm not going to add that into this PR because it could change downstream results, but am adding a TODO in the code and will file an issue in github.

tfenne commented on src/java/picard/analysis/AdapterUtil.java:
```diff
@@ -0,0 +1,105 @@
+import java.util.List;
+import java.util.Set;
+
+public class AdapterUtil {
```
Could you add a little javadoc to explain what this does and how it's different from `ClippingUtility` that looks vaguely similar?  I.e. that it's trying to match the starts of reads to any part of the adapter, vs. finding the start of the adapter anywhere in the read.

tfenne replied:
Looping back here, now that I see how this is used.  Why not make this class stateful instead of static?  I.e. turn `prepareAdapterSequences` into a constructor, hide all the `byte[]` and `byte[][]` stuff inside the class, and have the test method become just `public boolean isAdapterSequence(final byte[] read)`.

tfenne commented on src/java/picard/analysis/ChimeraUtil.java:
```diff
@@ -0,0 +1,70 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2015 The Broad Institute
```
2016

tfenne commented on src/java/picard/analysis/AdapterUtil.java:
```diff
@@ -0,0 +1,105 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2015 The Broad Institute
```
2016

tfenne commented on src/java/picard/analysis/ChimeraUtil.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.Set;
+
+public class ChimeraUtil {
+
```
Channeling Nils - extra newline here.

tfenne commented on src/java/picard/analysis/ChimeraUtil.java:
```diff
@@ -0,0 +1,70 @@
+     * @return true if this record is part of a chimeric read pair, false otherwise
+     */
+    public static boolean isChimeric(final SAMRecord rec, final int maxInsertSize, final Set<PairOrientation> expectedOrientations) {
+        return (Math.abs(rec.getInferredInsertSize()) > maxInsertSize ||
```
I think this method should either:
1. start with `if (rec.getReadUnmappedFlag() || !rec.getReadPairedFlag() || rec.getMateUnmappedFlag()) return false`
2. or make it explicit in the javaodoc that it's only legitimate to call this method on a read that is paired with both it and the mate mapped

tfenne commented on src/java/picard/analysis/ChimeraUtil.java:
```diff
@@ -0,0 +1,70 @@
+     * @param expectedOrientations   set of orientations that are not chimeric
+     * @return true if this pair is chimeric, false otherwise
+     */
+    public static boolean isChimeric(final SAMRecord r1, final SAMRecord r2, final int maxInsertSize, final Set<PairOrientation> expectedOrientations) {
```
Similarly, I think this should either test that the two reads are mapped, or explicitly state that the read must be mapped.

tfenne commented on src/java/picard/analysis/CollectAlignmentSummaryMetrics.java:
```diff
@@ -98,20 +97,13 @@
 
     @Option(doc="Paired-end reads that do not have this expected orientation will be considered chimeric.")
-    public PairOrientation EXPECTED_PAIR_ORIENTATION = PairOrientation.FR;
+    public Set<PairOrientation> EXPECTED_PAIR_ORIENTATIONS = ChimeraUtil.DEFAULT_EXPECTED_ORIENTATIONS;
```
To be safe, this should probably copy-construct the set of orientations.  I believe EnumSet is mutable, and that CommandLineProgram will mutate it!

tfenne commented on src/java/picard/analysis/CollectWgsMetricsFromQuerySorted.java:
```diff
@@ -195,10 +212,21 @@ private void calculateMetricsForRead(final QuerySortedReadPairIteratorUtil.ReadP
         if (isNoiseRead(pairToAnalyze.read1)) metrics.metrics.PF_NOISE_READS++;
-        if (!pairToAnalyze.read1.getReadUnmappedFlag()) metrics.metrics.PF_READS_ALIGNED++;
+
+        if (pairToAnalyze.read1.getReadUnmappedFlag()) {
```
Could be re-written as the following if you prefer:
```java
if (!pairToAnalyze.read1.getReadUnmappedFlag()) metrics.metrics.PF_READS_ALIGNED++;
else if (isAdapterRead(pairToAnalyze.read1)) metrics.metrics.PF_ADAPTER_READS++;
```

tfenne commented:
Looks good @eitanbanks - a bunch of suggestions from me, but nothing critical.

eitanbanks commented:
Reviews addressed.  Thanks, @tfenne.
Squashing and merging.
